Mini Cart Integration

hiro-iFactory opened the pull request:
- Initial commit.

hiro-iFactory committed:
Mini Cart Integration
- Initial commit.

ghost commented:
Hi all,
@AsifAmin  @gvanimurthy  @qld-gov-au/qld-online-development-delivery 
could you add this to our code review's we do.

regards,

@william-dutton-dsiti

william-dutton-dsiti commented on src/app/services/category.service.js:
```diff
@@ -97,6 +98,12 @@ class CategoryService {
     formatAutocompleteQuery(resourceId, filters, targetField) {
+
+        if ( ! targetField || targetField === '') {
+            console.log('targetField is missing');
```
should we log all the time?

hiro-iFactory replied:
I think I noticed [formatQuery() logs when missing resourceId ](https://github.com/qld-gov-au/qsa-archive-search/blob/caec3f5ea7b489037293b1125102048a9ca1f035/src/app/services/category.service.js#L52) so I must have assumed to do the same.

I am going to amend that part as logging only if a flag is `true`. The value of the flag is going to come from the new service I am mentioning below.

william-dutton-dsiti commented on src/app/providers/datatables.provider.js:
```diff
@@ -63,48 +63,87 @@ class DataTablesProvider {
                             <td colspan="${index.primary.length}">`;
 
         //--As a part of order online button--
+        const redirectUrl = 'https://test.smartservice.qld.gov.au/services/test/prodi/products';
```
how does this work between dev and prd?

hiro-iFactory replied:
I have amended that part by providing a new service called "EnvService" so that I can avoid hard-coding the endpoint. But I have not made the pull request with the latest code, yet, because there is something I have to confirm for something else.

ganapavz commented on src/app/templates/app.aside.html:
```diff
@@ -1,10 +1,31 @@
+        <div class="ssq-minicart-cards">
+            <h3>Cards accepted</h3>
+            <ul>
+                <li><img src="https://test.smartservice.qld.gov.au/payment/minicart/visa.png" alt="Visa"/></li>
```
would html hardcoding also be handled?

hiro-iFactory replied:
Yes. That part has been amended in the way mentioned [here](https://github.com/qld-gov-au/qsa-archive-search/pull/3#discussion_r143388164). The file is not going to have `https://test.smartservice.qld.gov.au` hard-coded as follows:

```html
(Omitting)
            <div class="ssq-minicart-submit">
                <input type="hidden" id="ssq-cart-contents" name="ssq-cart-contents" value=""/>
                <a href="{{vm.EnvService.smartServiceUrlPrefix}}/payment/cart/checkout"><img id="ssq_minicart_checkout"
                                                                  src="{{vm.EnvService.smartServiceUrlPrefix}}/payment/minicart/btn-checkout.png"
                                                                  alt="Checkout"/></a>
                <a href="{{vm.EnvService.smartServiceUrlPrefix}}/payment/cart/view"><img id="ssq_minicart_cart"
                                                              src="{{vm.EnvService.smartServiceUrlPrefix}}/payment/minicart/btn-cart.png"
                                                              alt="Edit cart"/></a>
(Omitting)
        <div class="ssq-minicart-cards">
            <h3>Cards accepted</h3>
            <ul>
                <li><img src="{{vm.EnvService.smartServiceUrlPrefix}}/payment/minicart/visa.png" alt="Visa"/></li>
                <li><img src="{{vm.EnvService.smartServiceUrlPrefix}}/payment/minicart/mastercard.png" alt="MasterCard"/></li>
(Omitting)
```

asifaminb commented on src/app/providers/datatables.provider.js:
```diff
@@ -63,48 +63,87 @@ class DataTablesProvider {
+        Price: <span ng-bind="vm.price"></span>
+      </fieldset>
+
+      <input type="hidden" name="attribute1" value="${attribute1}" />
```
I think using a **Directive** would be much better in this case?
Directives are **easy to debug and test**

hiro-iFactory replied:
I amended the code in the way already happening, but are you suggesting us to refactor this method so that we don't see HTML Tags here?

asifaminb replied:
Yes I think with Angular it is good if we have directives to encapsulate functions (HTML, js , css)
This helps to debug easily , add new functionality and writing test cases.

But if that is difficult to separate then we can leave as it is.

asifaminb commented on src/app/providers/datatables.provider.js:
```diff
@@ -151,12 +191,26 @@ class DataTablesProvider {
-                angular.element(row.node()).after(this.formatExtraInfo(index, row.node(), row.data()));
-                tr.addClass('shown');
+                let extraInfo = this.formatExtraInfo(index, row.node(), row.data());
+                angular.element(tr).injector().invoke(function($compile) {
```
What we are doing here ? 
A single line comment would help any dev jumping into this.

hiro-iFactory replied:
I have added the comment as follows:
```
let extraInfo = this.formatExtraInfo(EnvService, index, row.node(), row.data());
// Embed "extraInfo" after digesting the directives within (They are displayed as is otherwise)
angular.element(tr).injector().invoke(function($compile) {
```

hiro-iFactory committed:
Mini Cart Integration
- Enabled to configure environemnt dependant data (URL for the smart service etc).
- Included the product data for Live Environment.
- Changed the way to calculate the product price.

hiro-iFactory commented:
@asifamingov

So in our repository, we have committed the latest code having the amendments mentioned in this page except [the one to avoid HTML Tags within a method](https://github.com/qld-gov-au/qsa-archive-search/pull/3#discussion_r143598698) (as it is quite hard and we are not sure we can make it happen quickly).

And we would like the latest code to be reviewed, but GitHub does not let us create another Pull Request when there is an outstanding one. What is the best way to let you show it? Does GitHub let us create another request if we close this request?

asifaminb commented:
@hiro-iFactory 

Sorry for the late reply.
Sure that is just an improvement in the code if time permits.

- if you have made any drastic changes then I think it would be good to close this pull request and create a new pull request.

- If the changes are some minor enhancements then I think if you commit to the same branch then we can see the changes.

Please have a look and let me know if you have any other questions.

hiro-iFactory committed:
Mini Cart Integration
- Relocated the path for "category.json" to enable to load the file.
